src/index.js

```diff
@@ -616,58 +616,120 @@ class Service extends AdapterService {
-        }
-        q.upsertGraphAndFetch(data, this.upsertGraphOptions);
-      } else if (allowedInsert) {
-        q.allowGraph(allowedInsert);
-        q.insertGraph(data, this.insertGraphOptions);
-      } else {
-        q.insert(data, this.id);
-      }
-
-      return q
-        .then(row => {
-          if (params.query && params.query.$noSelect) { return data; }
+  _getCreatedRecords (insertResults, inputData, params) {
+    if (params.query && params.query.$noSelect) {
+      return inputData;
+    }
+    if (!Array.isArray(insertResults)) {
+      insertResults = [insertResults];
+    }
 
-          let id;
+    const findQuery = Object.assign({ $and: [] }, params.query);
+    const idsQueries = [];
 
-          if (Array.isArray(this.id)) {
-            id = [];
+    if (Array.isArray(this.id)) {
+      for (const insertResult of insertResults) {
+        const ids = [];
 
-            for (const idKey of this.id) {
-              id.push(row && row[idKey] ? row[idKey] : data[idKey]);
-            }
+        for (const idKey of this.id) {
+          if (idKey in insertResult) {
+            ids.push(insertResult[idKey]);
           } else {
-            id = row && row[this.id] ? row[this.id] : data[this.id];
+            return inputData;
           }
+        }
+        idsQueries.push(this.getIdsQuery(ids));
+      }
+    } else {
+      const ids = [];
+      for (const insertResult of insertResults) {
+        if (!(typeof insertResult === 'object' && insertResult !== null)) {
+          ids.push(insertResult);
+          continue;
+        }
 
-          if (!id || (Array.isArray(id) && !id.length)) { return data; }
+        if (this.id in insertResult) {
+          ids.push(insertResult[this.id]);
+        } else {
+          return inputData;
+        }
+      }
+      idsQueries.push(this.getIdsQuery(null, ids));
+    }
 
-          return this._get(id, params);
-        })
-        .catch(errorHandler);
-    };
+    if (idsQueries.length > 1) {
+      findQuery.$and.push({ $or: idsQueries });
+    } else {
+      findQuery.$and = findQuery.$and.concat(idsQueries);
+    }
 
-    if (Array.isArray(data)) {
-      return Promise.all(data.map(current => create(current, params))).then(this._commitTransaction(transaction), this._rollbackTransaction(transaction));
+    return this._find(Object.assign({}, params, { query: findQuery }))
+      .then(page => {
+        const records = page.data || page;
+        if (Array.isArray(inputData)) {
+          return records;
+        }
+        return records[0];
+      });
+  }
+
+  /**
+   * @param data
+   * @param params
+   * @returns {Promise<Object|(Object|string|number)[]>}
+   * @private
+   */
+  _batchInsert (data, params) {
+    const { dialect } = this.Model.knex().client;
+    const query = this._createQuery(params);
+
+    // .returning() is not supported by SQLite and MySQL
+    if (!(dialect === 'sqlite3' || dialect === 'mysql' || dialect === 'mysql2')) {
```

Comment:
The returning method is not supported on Amazon Redshift. how is it handled there? RedShift dialect is the same as PostgreSQL.

Reply:
Fixed. I used the dialect name from [here](https://github.com/knex/knex/blob/f953ff4f8f632b9d4e967e6072ac99a3da1e5fbb/lib/dialects/redshift/index.js#L44)